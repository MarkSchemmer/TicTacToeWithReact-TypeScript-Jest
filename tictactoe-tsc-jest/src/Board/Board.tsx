import * as React from 'react'
import Square from '../Square/Square'
import * as _ from '../Helpers/helpers'
interface IProps  {
    squares : Array<Array<_.Square>>
}

class Board extends React.Component<IProps> {
    render() {
        const { squares } = this.props
        return (
            <div>
                { squares.map((level) => level.map((sublevel) => {
                        return <Square
                                val={sublevel.val} 
                                coordinates={sublevel.coordinates}  /> 
                }))}          
            </div>
        )
    }
}

export default Board 