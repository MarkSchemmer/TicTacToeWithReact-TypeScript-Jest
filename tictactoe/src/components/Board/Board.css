
.row { 
    width:400px;
    height: 100px;
    /* border: 1px solid black; */
    margin: auto;
}


.column { 
    width: 130px;
    height: 100px;
    border : 1px solid black;
    display: inline-block;
}