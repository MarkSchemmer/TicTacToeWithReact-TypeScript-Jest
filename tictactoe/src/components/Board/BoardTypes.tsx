import React, { Component } from 'react'


interface ISquareAttributes {
    id : number, 
    coor : Array<number> 
    val : string | null 
}

interface IBoardWithMove {
    history : Array<Array<ISquareAttributes>>,
    Move : Array<Array<number>> | null,
    WhomMoved : string | null 
}


export class SquareAttributes implements ISquareAttributes {
    public id : number; 
    public coor : Array<number>;
    public val : string | null;

    constructor(id : number, coor : Array<number>, val : string | null = null){
        this.id = id
        this.coor = coor 
        this.val = val 
    }
}

export class BoardWithMoves implements IBoardWithMove {
    
    public history : Array<Array<SquareAttributes>>

    public Move : Array<Array<number>>

    public WhomMoved : string | null 
    
    constructor(history:Array<Array<SquareAttributes>>,
         Move : Array<Array<number>>, 
         WhomMoved : string | null = null ){
        this.history = history
        this.Move = Move 
        this.WhomMoved = WhomMoved
    }
}