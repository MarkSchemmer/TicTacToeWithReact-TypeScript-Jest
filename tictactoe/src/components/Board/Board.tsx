import React, { Component } from 'react'
import History from '../History/History'
import Square from '../Square/Square'
import './Board.css'

import * as api from '../../API/api'

import  * as BoardAttributes from './BoardTypes'

interface IProps {

}

interface IState {
    name : string, 
    board : Array<BoardAttributes.BoardWithMoves>,
    turn : number,
    IsWinner : number | null,
}

class Board extends Component<IProps, IState> {


    public state : IState = {
        name : 'Mark',
        board : [api.genBoard()],
        turn : 0,
        IsWinner : null, 
    }


    public onClick = (coor:Array<number>) => {
        const [x,y] = coor

        if(this.state.board[this.state.turn].history[x][y].val !== null) return  

        if(this.state.IsWinner !== null) return 

        // copy of history
        let copyOfHistory = this.state.board[this.state.turn].history

        // copy of Move 
        let copyOfMove = this.state.board[this.state.turn].Move 
        copyOfMove = [...copyOfMove, [x,y]]

        let whosMoving
    

        if(this.state.turn%2==0)
            whosMoving = 'X'
        else 
            whosMoving = 'O'

        copyOfHistory[x][y].val = whosMoving

        let newCopy = new BoardAttributes.BoardWithMoves(copyOfHistory, copyOfMove, whosMoving)

        this.setState((prevState) => {
            return {
                turn : prevState.turn+1,
                board : [...this.state.board, newCopy]
            }
        }, () => this.checkForWinner())
    }
 

    public checkForWinner = () => {
        console.log('the board: ',this.state.board)
        const hasWinner = api.IsWinner(this.state.board[this.state.turn].history)
        let whichOption = null 

        if(hasWinner===api.OPTIONS.XWIN)
            whichOption = api.OPTIONS.XWIN
        else if (hasWinner === api.OPTIONS.XLOSE) 
              whichOption = api.OPTIONS.XLOSE
        else if (hasWinner === api.OPTIONS.TIE)
                whichOption = api.OPTIONS.TIE

     if (whichOption)
        this.setState( { IsWinner : whichOption }, () => {
            this.setWinner()
        }) 
    }

    public setWinner = () => {
        if(this.state.IsWinner !== null){

            if(this.state.IsWinner===api.OPTIONS.TIE){
                return (
                    <React.Fragment>
                        TIE GAME!
                    </React.Fragment>
                )
            } else {
                if(this.state.IsWinner===api.OPTIONS.XWIN){
                    return (
                        <React.Fragment>
                            Player X You Win!
                        </React.Fragment>
                    )
                } else {
                   return ( <React.Fragment>
                        Player O You Win!
                   </React.Fragment> )
                }
            }

        } else { 
            return (
                <React.Fragment>
                    Player { this.state.turn % 2 === 0 ? <strong>{'X'}</strong> : <strong>{'O'}</strong> } your move!
                </React.Fragment>
            )
        }
    }

    public render () {
        return (
            <React.Fragment>
                <div className="main">
                <History history={this.state.board} /> 
                  <div className="greetings">Greetings { this.state.name }</div>
                  <div className="whoms-turn">  { this.setWinner() } </div>
                    <div className="board">
                            { this.state.board[this.state.turn].history.map((x:any) => 
                               <div className="row" key={ Date.now() * (Math.random() * 100) }>
                                    { x.map((z:any) => <Square key={z.id} {...z} 
                                        onClick={this.onClick.bind(this)} /> )}
                               </div> )}
                    </div>
                </div>
            </React.Fragment>
        )
    }
}

export default Board 

