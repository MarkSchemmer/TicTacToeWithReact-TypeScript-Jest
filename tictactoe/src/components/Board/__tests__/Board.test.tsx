import * as React from 'react'
import * as enzyme from 'enzyme'
import  Adapter from 'enzyme-adapter-react-16'
import renderer from 'react-test-renderer'
import Board from '../Board'


/*

    Some quick notes on Shallow, Mount, Render : 
        Shallow:  rendering is testing the component as a unit, also this shallow rendering does 
        render any child elements...
        
        Mount: is full dom rendering of component and child components, and remember that mount actually 
            mounts the component in the DOM and any changes will affect other children... 

        Render: I guess it renders the component to static html but uses the cheerio api for traversing 
*/

enzyme.configure({ adapter: new Adapter() });

describe('testing that <Board /> has greetings tag with my name', () => {

    let greetings 
    beforeEach(() => {
        greetings = enzyme.shallow( <Board /> )
    })

    it(' Should be true! ', () => {
            expect(greetings.find('.greetings').text()).toEqual('Greetings Mark')
    }) 

    it(' should be false ', () => {
        let expected = greetings.find('.greetings').text() === 'Mark Greetingsw'
        expect(expected).toEqual(false)
    })

    // board renders with 3 rows and 3 columns...  

    it( ' has 3 rows ? ', () => {  
        expect(greetings.find('.row')).toHaveLength(3)
    })


    // board should have 9 columns... 

    it(' should have 9 columns?', () => {
        expect(greetings.find('.column')).toHaveLength(9)
    })
    
    // has proper state 

    it(' has state items Mark and Board', () => {
        const wrapper = enzyme.mount(<Board /> )

        expect(wrapper.state().name).toEqual('Mark')
        expect(wrapper.state().board.length).toEqual(3)
    })

    // setup snapshot testing...   
    // not using the enzyme  snapshot... 

    it('Snapshot test for <Board /> ', () => {
        const tree = renderer.create(<Board />).toJSON() 
        expect(tree).toMatchSnapshot() 
    })

})



