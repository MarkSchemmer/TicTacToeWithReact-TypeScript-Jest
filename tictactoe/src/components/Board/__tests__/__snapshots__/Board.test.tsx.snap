// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test for <Board />  1`] = `
<div
  className="main"
>
  <div
    className="greetings"
  >
    Greetings 
    Mark
  </div>
  <div
    className="whoms-turn"
  >
      
    Player 
    <strong>
      X
    </strong>
     your move!
     
  </div>
  <div
    className="board"
  >
    <div
      className="row"
    >
      <span
        className="column"
        onClick={[Function]}
      >
        <div
          className="square-val"
        />
      </span>
      <span
        className="column"
        onClick={[Function]}
      >
        <div
          className="square-val"
        />
      </span>
      <span
        className="column"
        onClick={[Function]}
      >
        <div
          className="square-val"
        />
      </span>
    </div>
    <div
      className="row"
    >
      <span
        className="column"
        onClick={[Function]}
      >
        <div
          className="square-val"
        />
      </span>
      <span
        className="column"
        onClick={[Function]}
      >
        <div
          className="square-val"
        />
      </span>
      <span
        className="column"
        onClick={[Function]}
      >
        <div
          className="square-val"
        />
      </span>
    </div>
    <div
      className="row"
    >
      <span
        className="column"
        onClick={[Function]}
      >
        <div
          className="square-val"
        />
      </span>
      <span
        className="column"
        onClick={[Function]}
      >
        <div
          className="square-val"
        />
      </span>
      <span
        className="column"
        onClick={[Function]}
      >
        <div
          className="square-val"
        />
      </span>
    </div>
  </div>
</div>
`;
